src/script/api/script_tilelist.cpp

```diff
@@ -123,11 +111,7 @@ ScriptTileList_IndustryProducing::ScriptTileList_IndustryProducing(IndustryID in
 	if (i->neutral_station != nullptr && !_settings_game.station.serve_neutral_industries) return;
 
 	/* Check if this industry produces anything */
-	bool cargo_produces = false;
-	for (byte j = 0; j < lengthof(i->produced_cargo); j++) {
-		if (::IsValidCargoID(i->produced_cargo[j])) cargo_produces = true;
-	}
-	if (!cargo_produces) return;
+	if (std::all_of(std::begin(i->produced), std::end(i->produced), [](const auto &p) { return !::IsValidCargoID(p.cargo); })) return;
```

Comment:
```suggestion
	if (std::none_of(std::begin(i->produced), std::end(i->produced), [](const auto &p) { return ::IsValidCargoID(p.cargo); })) return;
```
Also, given similar codes get duplicated all around. Maybe a `HasCargo` and `HasAnyCargo` function in `CargoArray`, then a lot of the `std::all_of/any_of/none_of` variants can be "hidden" by a more descriptive bit of code.

Reply:
Yes, except it's not a CargoArray :)